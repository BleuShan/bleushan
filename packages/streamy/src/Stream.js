import {invariant} from './utils/invariant.js'
import {isArrayLike} from './utils/isArrayLike.js'
import {isString} from './utils/isString.js'
/* eslint-disable no-unused-vars */
import {typeOf} from './utils/typeOf.js'
import {IndexIterator} from './iterators/IndexIterator.js'
/* eslint-enable no-unused-vars */

export class Stream {
  #currentMemoizeResultId = -1
  #hasCompleted = false
  #isAsync
  #memoizedResults = new Map()
  #source
  #sourceIterator

  // eslint-disable-next-line no-undef
  get #iterator() {
    if (this.#sourceIterator == null) {
      if (this.#hasCompleted) {
        this.#sourceIterator = this.#memoizedResults.values()
      } else if (this.isAsync) {
        this.#sourceIterator = this.#source[Symbol.asyncIterator]()
      } else if (typeOf(this.#source[Symbol.iterator]) === 'function') {
        this.#sourceIterator = this.#source[Symbol.iterator]()
      } else {
        this.#sourceIterator = new IndexIterator(this.#source)
      }
    }

    return this.#sourceIterator
  }

  get isAsync() {
    return this.#isAsync
  }

  get hasCompleted() {
    return this.#hasCompleted
  }

  constructor(source) {
    invariant({
      condition: isArrayLike(source) || isString(source),
      message: 'source must be a string, an ArrayLike object or an iterable',
      errorType: TypeError
    })

    this.#source = source
    this.#isAsync = typeOf(this.#source[Symbol.asyncIterator]) === 'function'
  }

  [Symbol.asyncIterator]() {
    return this
  }

  [Symbol.iterator]() {
    invariant({
      condition: !this.isAsync || this.#hasCompleted,
      message: 'Cannot iterate synchronously over an uncompleted asynchronous stream',
      errorType: TypeError
    })

    return this
  }

  next(...input) {
    let promiseOrResult
    if (input.length) {
      promiseOrResult = this.#memoizedResultFor(...input)
    }

    if (promiseOrResult == null) {
      promiseOrResult = this.#iterator.next(...input)

      if (!this.#hasCompleted) {
        if (typeOf(promiseOrResult.then) === 'function') {
          return promiseOrResult.then((result) => this.#memoize(result, ...input))
        } else {
          return this.#memoize(promiseOrResult, ...input)
        }
      }
    }

    return promiseOrResult
  }

  // eslint-disable-next-line no-undef
  #memoize({value, done}, ...input) {
    this.#hasCompleted = done && value === undefined

    if (!this.#hasCompleted) {
      this.#memoizedResults.set(this.#memoizedResultKeyFrom(...input), value)
    } else {
      this.#sourceIterator = null
    }

    return {
      value,
      done: this.#hasCompleted
    }
  }

  // eslint-disable-next-line no-undef
  #memoizedResultFor(...input) {
    const key = this.#memoizedResultKeyFor(input)
    if (this.#memoizedResults.has(key)) {
      const value = this.#memoizedResults.get(key)

      return {
        value,
        done: false
      }
    }
  }

  // eslint-disable-next-line no-undef
  #memoizedResultKeyFor(input) {
    return JSON.stringify(input, (key, value) => (value === undefined ? 'undefined' : value))
  }

  // eslint-disable-next-line no-undef
  #memoizedResultKeyFrom(...input) {
    return input.length ? this.#memoizedResultKeyFor(input) : ++this.#currentMemoizeResultId
  }
}
