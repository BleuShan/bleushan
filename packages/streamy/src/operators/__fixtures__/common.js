export function toArray(iterable) {
  if (Array.isArray(iterable)) {
    return iterable
  }

  if (iterable.length != null) {
    return Array.from(iterable)
  }

  const result = []
  for (const value of iterable) {
    result.push(value)
  }
  return result
}

export async function toArrayAsync(iterable) {
  if (Array.isArray(iterable)) {
    return iterable
  }

  if (iterable.length != null) {
    return Array.from(iterable)
  }

  const result = []
  for await (const value of iterable) {
    result.push(value)
  }
  return result
}
