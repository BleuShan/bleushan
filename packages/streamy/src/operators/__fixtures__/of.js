export function* generator() {
  yield 1
  yield 2
  yield 3
}

export async function* asyncGenerator() {
  for (let i = 1; i < 10; i++) {
    const result = await new Promise((resolve) => {
      resolve(i * 2)
    })
    yield result
  }
}

export class AsyncIterable {
  #count
  constructor() {
    this.#count = 0
  }

  [Symbol.asyncIterator]() {
    return this
  }

  async next() {
    const result = {
      done: this.#count >= 10
    }

    if (!result.done) {
      result.value = ++this.#count
    }

    return result
  }

  reset() {
    this.#count = 0
  }
}

export async function iterateAsync(iterable) {
  const result = []
  for await (const value of iterable) {
    result.push(value)
  }
  return result
}

export function iterateSync(iterable) {
  const result = []
  for (const value of iterable) {
    result.push(value)
  }
  return result
}
