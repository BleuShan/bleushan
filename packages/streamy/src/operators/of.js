import {instanceOf} from '@bleushan/utils'
import Stream from '../Stream.js'

const isStream = instanceOf(Stream)

export default function of(...sources) {
  if (sources.length === 1) {
    const [source] = sources

    if (source != null && isStream(source)) {
      return source
    }
  }

  return new Stream(...sources)
}
