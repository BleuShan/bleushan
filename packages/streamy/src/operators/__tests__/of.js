import {typeTag, isArrayLike, isString, isFunction, instanceOf} from '@bleushan/utils/lib/cjs'
import {asyncGenerator, AsyncIterable, generator} from '../__fixtures__/of.js'
import {toArray, toArrayAsync} from '../__fixtures__/common.js'
import of from '../of.js'
import Stream from '../../Stream.js'

describe('the "of" operator', () => {
  describe('when invoked with', () => {
    let invoke
    let expectedResult
    let source
    let stream

    describe('a single', () => {
      beforeEach(async () => {
        const isGenerator =
          typeTag(source) === 'GeneratorFunction' || typeTag(source) === 'AsyncGeneratorFunction'

        if (isArrayLike(source) || isString(source) || isGenerator) {
          const iter = isGenerator ? source() : source
          if (isFunction(iter[Symbol.asyncIterator])) {
            expectedResult = await toArrayAsync(iter)
          } else {
            expectedResult = toArray(iter)
          }
        } else if (isFunction(source?.then)) {
          const value = await source
          expectedResult = [value]
        }

        invoke = () => {
          stream = isGenerator ? of(source()) : of(source)

          return stream
        }
      })

      describe.each([
        [[1, 2, 3]],
        [{0: 1, length: 1}],
        ['sast'],
        [new Set(['s'])],
        [generator],
        [asyncGenerator],
        [new AsyncIterable()],
        [Promise.resolve(1)],
        [of([1, 2, 3])]
      ])('valid value, like %o,', (value) => {
        beforeAll(() => {
          source = value
        })

        it('should not throw any error', () => {
          expect(invoke).not.toThrow()
        })

        describe('from which the result', () => {
          it('should be an instance of Stream', () => {
            expect(stream).toBeInstanceOf(Stream)
          })

          it('should yield the expected iteration result ', async () => {
            const result = await toArrayAsync(stream)
            expect(stream.completed).toEqual(true)
            expect(result).toEqual(expectedResult)
          })

          if (
            (isFunction(value[Symbol.asyncIterator]) ||
              isFunction(value.then) ||
              typeTag(value) === 'AsyncGeneratorFunction') &&
            !instanceOf(Stream, value)
          ) {
            it('should not allow synchronous traversal if the stream is completed', () => {
              expect(() => toArray(stream)).toThrowErrorMatchingSnapshot()
            })

            it('should not allow synchronous traversal if the stream is not completed', () => {
              expect(() => toArray(invoke())).toThrowErrorMatchingSnapshot()
            })
          }

          if (instanceOf(Stream, value)) {
            it('should be a noop if the value is a Stream', () => {
              expect(stream).toBe(value)
            })
          }
        })
      })

      describe.each([
        [
          {
            test: 's'
          }
        ],
        [null],
        [undefined],
        [NaN],
        [1],
        [BigInt(1)]
      ])('invalid value, like %o,', (value) => {
        beforeAll(() => {
          source = value
        })

        it('should throw the appropriate error', () => {
          expect(invoke).toThrowErrorMatchingSnapshot()
        })
      })
    })

    describe('multiple', () => {
      beforeEach(() => {
        expectedResult = toArray(source)

        invoke = () => {
          stream = of(...source)

          return stream
        }
      })

      describe.each([
        [1, 2, 3],
        [
          {0: 1, length: 1},
          {0: 2, length: 1},
          {0: 3, length: 1}
        ],
        ['a', 'b', 'c'],
        [Promise.resolve(1), Promise.resolve(2), Promise.resolve(3)],
        [NaN, null, undefined]
      ])('values, like %o, %o and %o,', (...values) => {
        beforeAll(() => {
          source = values
        })

        it('should not throw any error', () => {
          expect(invoke).not.toThrow()
        })

        describe('from which the result', () => {
          it('should be an instance of Stream', () => {
            expect(stream).toBeInstanceOf(Stream)
          })

          it('should yield the expected iteration result ', () => {
            const result = toArray(stream)
            expect(result).toEqual(expectedResult)
          })
        })
      })
    })
  })
})
