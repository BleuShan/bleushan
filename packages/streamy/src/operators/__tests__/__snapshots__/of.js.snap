// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the "of" operator when invoked with a single invalid value, like { test: 's' }, should throw the appropriate error 1`] = `"source must be a string, an iterable, an ArrayLike or PromiseLike object"`;

exports[`the "of" operator when invoked with a single invalid value, like 1, should throw the appropriate error 1`] = `"source must be a string, an iterable, an ArrayLike or PromiseLike object"`;

exports[`the "of" operator when invoked with a single invalid value, like 1n, should throw the appropriate error 1`] = `"source must be a string, an iterable, an ArrayLike or PromiseLike object"`;

exports[`the "of" operator when invoked with a single invalid value, like NaN, should throw the appropriate error 1`] = `"source must be a string, an iterable, an ArrayLike or PromiseLike object"`;

exports[`the "of" operator when invoked with a single invalid value, like null, should throw the appropriate error 1`] = `"source must be a string, an iterable, an ArrayLike or PromiseLike object"`;

exports[`the "of" operator when invoked with a single invalid value, like undefined, should throw the appropriate error 1`] = `"source must be a string, an iterable, an ArrayLike or PromiseLike object"`;

exports[`the "of" operator when invoked with a single valid value, like [AsyncGeneratorFunction: asyncGenerator] {
  [length]: 0,
  [name]: 'asyncGenerator',
  [prototype]: Object [AsyncGenerator] {}
}, from which the result should not allow synchronous traversal if the stream is completed 1`] = `"Cannot traverse an asynchronous stream synchronously"`;

exports[`the "of" operator when invoked with a single valid value, like [AsyncGeneratorFunction: asyncGenerator] {
  [length]: 0,
  [name]: 'asyncGenerator',
  [prototype]: Object [AsyncGenerator] {}
}, from which the result should not allow synchronous traversal if the stream is not completed 1`] = `"Cannot traverse an asynchronous stream synchronously"`;

exports[`the "of" operator when invoked with a single valid value, like AsyncIterable {}, from which the result should not allow synchronous traversal if the stream is completed 1`] = `"Cannot traverse an asynchronous stream synchronously"`;

exports[`the "of" operator when invoked with a single valid value, like AsyncIterable {}, from which the result should not allow synchronous traversal if the stream is not completed 1`] = `"Cannot traverse an asynchronous stream synchronously"`;

exports[`the "of" operator when invoked with a single valid value, like Promise { 1 }, from which the result should not allow synchronous traversal if the stream is completed 1`] = `"Cannot traverse an asynchronous stream synchronously"`;

exports[`the "of" operator when invoked with a single valid value, like Promise { 1 }, from which the result should not allow synchronous traversal if the stream is not completed 1`] = `"Cannot traverse an asynchronous stream synchronously"`;
